src/QuadForm/Quad/ZGenusRep.jl

```diff
@@ -0,0 +1,950 @@
+  # in `K^{rank(L)}`, and we then test a random sample of lines. By default,
+  # we test at most 10 lines, but the user can choose to change this parameter.
+  P = enumerate_lines(K, rank(L))
+
+  maxlines = min(rand_neigh, length(P))
+
+  result = typeof(L)[]
+
+  @vprintln :ZGenRep 3 "Try $(maxlines) random lines"
+
+  for i in 1:maxlines
+    vain[] > stop_after && break
+    w = matrix(QQ, 1, rank(L), ZZRingElem[lift(ZZ, k) for k in rand(P)])
+    a = numerator(only(w*form0*transpose(w)))
+    if !is_divisible_by(a, mqf)
+      vain[] += 1
+      continue
+    end
+
+    lifts = typeof(w)[]
+    if p == 2
+      bo = is_divisible_by(a, 8)
+      if even && !bo
+        if prime_dual(L, w, p) != L0
+          vain[] += 1
+          continue
+        end
+        make_admissible!(w, form0, m, K, a)
+        push!(lifts, w)
+      elseif even && bo
+        push!(lifts, w)
+      elseif !even && bo
+        if prime_dual(L, w*L0toL, p) != L0
+          vain[] += 1
+          continue
+        end
+        push!(lifts, w*L0toL)
+      else
+        if prime_dual(L, w*L0toL, p) == L0
+          push!(lifts, w*L0toL)
+        elseif prime_dual(L0, w, p) == L0
+          push!(lifts, w*L0toL)
+        else
+          push!(lifts, w*L0toL)
+          make_admissible!(w, form0, ZZ(8), K, a)
+          push!(lifts, w*L0toL)
+        end
+      end
+    else
+      !is_divisible_by(a, m) && make_admissible!(w, form0, m, K, a)
+      w = flag ? w*L0toL : w
+      push!(lifts, w)
+    end
+
+    for v in lifts
+      LL = lll(neighbour(L, v, p))
+      @hassert :ZGenRep 3 is_locally_isometric(LL, L, p)
+
+      keep = callback(LL)
+      if !keep
+        vain[] += 1
+        continue
+      end
+
+      vain[] = Int(0)
+      @vprintln :ZGenRep 3 "Keep an isometry class"
+      invLL = _invariants(LL)
+      if haskey(inv_dict, invLL)
+        push!(inv_dict[invLL], LL)
+      else
+        inv_dict[invLL] = ZZLat[LL]
+      end
+      push!(result, LL)
+
+      if save_partial
+        save_partial_lattice(LL, save_path)
+      end
+
+      if use_mass
+        s = automorphism_group_order(LL)
+        sub!(__mass, __mass, 1//s)
+        is_zero(__mass) && return result
+      end
+
+      length(result) == max && break
+    end
+  end
+  return result
+end
+
+###############################################################################
+#
+#  Invariants
+#
+###############################################################################
+
+# Resize `A` to only keep one representative for each isometry class
+# represented by the lattices in `A`.
+function _unique_iso_class!(A::Vector{ZZLat})
+  isempty(A) && return A
+  idxs = eachindex(A)
+  T = NTuple{2, Any}
+  it = iterate(idxs, (iterate(idxs)::T)[2])
+  count = 1
+  for x in Iterators.drop(A, 1)
+    if all(y -> !is_isometric(x, y), A[1:count])
+      A[it[1]] = x
+      count += 1
+      it = iterate(idxs, it[2])
+    end
+  end
+  resize!(A, count)::typeof(A)
+end
```

Comment:
I am having a hard time to read and understand this piece of code. Can you insert some comments as to what happens?
For instance `(iterate(idxs)::T)[2]` seems to be equivalent to `idxs[1]`?